### PR TITLE
add max n classical register

### PR DESCRIPTION
**Pull request by cqc-melf**
add new max n classical register predicate

### CODE REVIEW

File: tket/src/Predicates/Predicates.cpp [hunk 1]
```diff
@@ -525,6 +526,47 @@ std::string MaxNQubitsPredicate::to_string() const {
   return auto_name(*this) + "(" + std::to_string(n_qubits_) + ")";
 }
 
+bool MaxNClRegPredicate::verify(const Circuit& circ) const {
+  bit_vector_t all_bits = circ.all_bits();
+  std::map<std::string, unsigned> bits_map;
```
Comment (cqc-alec) [inline tket/src/Predicates/Predicates.cpp:531]: This doesn't need to be a map, does it? Can just be a set of string.

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: pytket/docs/changelog.rst [hunk 2]
```diff
@@ -1,6 +1,14 @@
 Changelog
 =========
 
+1.15.0 (unreleased)
+-------------------
+
+Minor new features:
+
+* Add new ``MaxNClRegPredicate`` that checks that there are at max n classical
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply pytket/docs/changelog.rst]: done in @cqc-melf
[requested changes](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)
[14fa39f](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)

File: pytket/pytket/backends/backendinfo.py [hunk 3]
```diff
@@ -322,5 +323,5 @@ def fully_connected_backendinfo(  # type: ignore
     :param \\**kwargs: All further arguments are passed to the BackendInfo constructor.
     """
     return BackendInfo(
-        name, device_name, version, FullyConnected(n_nodes), gate_set, **kwargs
+        name, device_name, version, FullyConnected(n_nodes), gate_set, None, **kwargs
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply pytket/pytket/backends/backendinfo.py]: Yes, done in @cqc-melf
[requested changes](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)
[14fa39f](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)

File: tket/test/src/test_Predicates.cpp [hunk 4]
```diff
@@ -354,6 +354,58 @@ SCENARIO("Test routing-related predicates' meet and implication") {
...
+
+    PredicatePtr con0 = std::make_shared<MaxNClRegPredicate>(0);
+    PredicatePtr con1 = std::make_shared<MaxNClRegPredicate>(1);
+    PredicatePtr con5 = std::make_shared<MaxNClRegPredicate>(5);
+
+    REQUIRE(con0->verify(circ));
+    REQUIRE(con1->verify(circ));
+    REQUIRE(con5->verify(circ));
+  }
+  GIVEN("no classical reg") {
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply tket/test/src/test_Predicates.cpp]: Fixed in @cqc-melf
[requested changes](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)
[14fa39f](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)

File: tket/test/src/test_Predicates.cpp [hunk 5]
```diff
@@ -354,6 +354,58 @@ SCENARIO("Test routing-related predicates' meet and implication") {
...
+    MaxNClRegPredicate con0 = MaxNClRegPredicate(0);
+    PredicatePtr con3 = std::make_shared<MaxNClRegPredicate>(3);
+    PredicatePtr con5 = std::make_shared<MaxNClRegPredicate>(5);
+    MaxNClRegPredicate con5b = MaxNClRegPredicate(5);
+    REQUIRE(con0.implies(*con5));
+    PredicatePtr con3b = con5->meet(*con3);
+    REQUIRE(con3b->implies(*con5));
+    PredicatePtr con3c = con5b.meet(*con3);
+    REQUIRE(con3b->implies(*con5));
+    con0.to_string();
```
[the remark this answers is not in the document]
Comment (cqc-melf) [reply tket/test/src/test_Predicates.cpp]: Yes, I have just asserted that expected value, done in @cqc-melf
[requested changes](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)
[14fa39f](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)

File: tket/src/Predicates/Predicates.cpp (hunk 1 above)
Comment (cqc-melf) [reply tket/src/Predicates/Predicates.cpp]: I have replaced that with a set and removed the else case as well. Done in @cqc-melf
[requested changes](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08)
[14fa39f](https://github.com/CQCL/tket/pull/822/commits/14fa39fb728ff26eb74db4d1c34d5eb86cf21c08) and @cqc-melf
[bit sets rename](https://github.com/CQCL/tket/pull/822/commits/7beba097857da497b3833f4f9eea03e8fd1a17b6)
[7beba09](https://github.com/CQCL/tket/pull/822/commits/7beba097857da497b3833f4f9eea03e8fd1a17b6)

File: tket/src/Predicates/Predicates.cpp [hunk 6]
```diff
@@ -525,6 +526,45 @@ std::string MaxNQubitsPredicate::to_string() const {
...
 }
 
+bool MaxNClRegPredicate::verify(const Circuit& circ) const {
+  bit_vector_t all_bits = circ.all_bits();
+  std::set<std::string> bits_set;
+  for (Bit bit : all_bits) {
+    auto it = bits_set.find(bit.reg_name());
+    if (it == bits_set.end()) {
+      bits_set.insert(bit.reg_name());
+    }
```
Comment (cqc-alec) [inline tket/src/Predicates/Predicates.cpp:536]:
```suggestion
    bits_set.insert(bit.reg_name());
```
No need to check if it's there already, this is automatic for sets.

**Review by cqc-alec — CHANGES_REQUESTED** [review]

File: tket/src/Predicates/Predicates.cpp (hunk 6 above)
Comment (cqc-melf) [reply tket/src/Predicates/Predicates.cpp]: Thank you, good point!

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**